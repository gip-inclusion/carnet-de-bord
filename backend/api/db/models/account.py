from datetime import datetime
from uuid import UUID

from pydantic import BaseModel

from api.db.models.role import RoleEnum


class AccountDB(BaseModel):
    id: UUID
    username: str
    type: RoleEnum
    access_key: UUID | None
    access_key_date: datetime | None
    last_login: datetime | None
    admin_id: UUID | None
    manager_id: UUID | None
    orientation_manager_id: UUID | None
    admin_structure_id: UUID | None
    professional_id: UUID | None
    beneficiary_id: UUID | None
    confirmed: bool = False
    onboarding_done: bool = False
    created_at: datetime
    updated_at: datetime


class AccountDBWithAccessKey(AccountDB):
    access_key: UUID
    access_key_date: datetime


class AccountInfo(BaseModel):
    account_id: UUID
    firstname: str
    lastname: str
    email: str | None
