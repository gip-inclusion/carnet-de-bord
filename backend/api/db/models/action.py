from datetime import datetime
from uuid import UUID

from pydantic import BaseModel


class Action(BaseModel):
    id: UUID
    action: str
    target_id: UUID
    status: str
    creator_id: UUID
    created_at: datetime
    updated_at: datetime
    initial_id: str | None
