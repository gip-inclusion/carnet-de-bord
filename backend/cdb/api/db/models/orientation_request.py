from datetime import datetime
from uuid import UUID

from pydantic import BaseModel


class OrientationRequest(BaseModel):
    id: UUID
    created_at: datetime
    updated_at: datetime
    decided_at: datetime | None
    beneficiary_id: UUID
    requestor_account_id: UUID
    requested_orientation_system_id: UUID | None
    decided_orientation_system_id: UUID | None
    status: str
    reason: str | None
