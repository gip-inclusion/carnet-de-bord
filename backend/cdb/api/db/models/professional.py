from datetime import datetime
from uuid import UUID

from pydantic import BaseModel


class ProfessionalInsert(BaseModel):
    structure_id: UUID
    email: str | None
    lastname: str
    firstname: str
    position: str | None
    mobile_number: str | None


class Professional(ProfessionalInsert):
    id: UUID
    account_id: UUID | None
    created_at: datetime
    updated_at: datetime
