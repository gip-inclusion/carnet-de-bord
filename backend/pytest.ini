[pytest]
asyncio_mode = auto
testpaths = "tests"
addopts = --strict-markers --strict-config
markers =
    graphql: Used for tests that rely on GraphQL. Those tests actually write to the database, which needs to be re-seeded. Seeding being expensive, prefer the use of SQL.
#log_cli = true
#log_level = INFO
