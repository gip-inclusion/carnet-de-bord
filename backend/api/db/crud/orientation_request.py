import logging
from uuid import UUID

from asyncpg import Record
from asyncpg.connection import Connection

from api.db.models.orientation_request import OrientationRequest

logger = logging.getLogger(__name__)


async def get_orientation_request_by_id(
    connection: Connection, orientation_request_id: UUID
) -> OrientationRequest | None:
    orientation_request: Record | None = await connection.fetchrow(
        "SELECT * FROM orientation_request WHERE id = $1", orientation_request_id
    )

    if orientation_request:
        return OrientationRequest.parse_obj(orientation_request)
