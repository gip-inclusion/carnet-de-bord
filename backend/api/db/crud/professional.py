import logging

from asyncpg import Record
from asyncpg.connection import Connection

from api.db.models.professional import Professional, ProfessionalInsert

logger = logging.getLogger(__name__)


async def get_professional_with_query(
    connection: Connection, query: str, *args
) -> Professional | None:
    async with connection.transaction():

        professional_record: Record | None = await connection.fetchrow(
            "SELECT public.professional.*, public.account.id as account_id "
            "FROM public.professional "
            "LEFT JOIN public.account "
            "ON public.account.professional_id = public.professional.id " + query,
            *args,
        )

        if professional_record:
            return Professional.parse_obj(professional_record)


async def get_professional_by_email(
    connection: Connection, email: str
) -> Professional | None:
    return await get_professional_with_query(
        connection,
        "WHERE LOWER(public.professional.email) = LOWER($1)",
        email,
    )


async def insert_professional(
    connection: Connection, professional_insert: ProfessionalInsert
) -> Professional | None:

    record = await connection.fetchrow(
        """
        INSERT INTO public.professional (structure_id, email, lastname, firstname, position, mobile_number)
        VALUES ($1, $2, $3, $4, $5, $6) returning id, structure_id, email, lastname, firstname, position, mobile_number, created_at, updated_at
        """,
        *professional_insert.dict().values(),
    )

    if record:
        return await parse_professional_from_record(record)


async def parse_professional_from_record(record: Record) -> Professional:
    return Professional.parse_obj(record)
