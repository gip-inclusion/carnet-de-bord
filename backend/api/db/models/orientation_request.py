from datetime import datetime
from uuid import UUID

from pydantic import BaseModel, Field

from api.db.models.orientation_type import OrientationType


class OrientationRequest(BaseModel):
    id: UUID
    created_at: datetime
    updated_at: datetime
    decided_at: datetime | None
    beneficiary_id: UUID
    requestor_account_id: UUID
    requested_orientation_type_id: OrientationType | None
    decided_orientation_type_id: OrientationType | None
    status: str
    reason: str | None
