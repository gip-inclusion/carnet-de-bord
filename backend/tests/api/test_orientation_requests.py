from unittest import mock

import pytest
from httpx import AsyncClient

from cdb.api.db.models.orientation_request import OrientationRequest
from cdb.api.db.models.professional import Professional
from tests.utils.approvaltests import verify

pytestmark = pytest.mark.graphql

DENY_ORIENTATION_ENDPOINT_PATH = "/v1/orientation_requests/deny"


@mock.patch("cdb.api.core.emails.send_mail")
async def test_deny_orientation_request_email(
    mock_send_email: mock.Mock,
    test_client: AsyncClient,
    professional_edith_orial: Professional,
    orientation_request_jennings_dee: OrientationRequest,
    giulia_diaby_jwt: str,
):
    response = await test_client.post(
        DENY_ORIENTATION_ENDPOINT_PATH,
        json={
            "orientation_request_id": orientation_request_jennings_dee.id,
        },
        headers={"Authorization": "Bearer " + giulia_diaby_jwt},
    )
    print(response.json())
    assert response.status_code == 200
    assert mock_send_email.call_count == 1

    email_new_referent = mock_send_email.call_args_list[0]
    assert email_new_referent.kwargs["to"] == professional_edith_orial.email
    assert email_new_referent.kwargs["subject"] == "Maintien de lâ€™accompagnement"
    verify(email_new_referent.kwargs["message"], extension=".html")
