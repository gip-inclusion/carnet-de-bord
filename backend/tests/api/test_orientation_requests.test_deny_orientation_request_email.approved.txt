{'email_new_referent': '<!DOCTYPE html>\n<html lang="fr">\n\t<head>\n\t\t<meta charset="UTF-8" />\n\t\t<meta http-equiv="X-UA-Compatible" content="IE=edge" />\n\t\t<meta name="viewport" content="width=device-width, initial-scale=1.0" />\n\t\t<title>Maintien de l’accompagnement</title>\n\t</head>\n\t<body>\n\t\t<h1>Maintien de l’accompagnement</h1>\n\t\t<strong>Suite à votre demande de réorientation, il a été décidé de maintenir l’accompagnement</strong>\n\t\t<p>Bénéficiaire&nbsp;: <strong>Dee Jennings</strong></p>\n\n\t\t<p>\n\t\t\tPour plus de renseignements, veuillez contacter votre mandataire.\n\t\t</p>\n\n\t\t<p>\n\t\t\tPour accéder à Carnet de bord, veuillez cliquer sur le lien\n\t\t\tci-dessous&nbsp;:\n\t\t</p>\n\t\t<p style="padding-left: 20%">\n\t\t\t<a\n\t\t\t\trel="nofollow"\n\t\t\t\thref="http://localhost:3001"\n\t\t\t\tstyle="\n\t\t\t\t\tbackground-color: #6a20ae;\n\t\t\t\t\tfont-size: 14px;\n\t\t\t\t\tfont-family: Helvetica, Arial, sans-serif;\n\t\t\t\t\tfont-weight: bold;\n\t\t\t\t\ttext-decoration: none;\n\t\t\t\t\tpadding: 8px 10px;\n\t\t\t\t\tcolor: #ffffff;\n\t\t\t\t\tborder-radius: 5px;\n\t\t\t\t\tdisplay: inline-block;\n\t\t\t\t"\n\t\t\t>\n\t\t\t\t<span>Accédez à Carnet de bord</span>\n\t\t\t</a>\n\t\t</p>\n\t\t<p>L\'équipe Carnet de bord</p>\n\t</body>\n</html>'}
